src/tunnel.cc

```diff
@@ -318,6 +318,8 @@ void
 TunnelStateData::serverClosed()
 {
     server.noteClosure();
+
+    retryOrBail(__FUNCTION__);
```

Comment:
I agree that this fixes memory leaks. I think retries should not possible due to TunnelStateData::retriable [flag](https://github.com/squid-cache/squid/pull/489#discussion_r780835913) protection.